### PR TITLE
[FLINK-4928] [yarn] Implement FLIP-6 YARN Application Master Runner

### PR DESCRIPTION
**Pull request by shuai-xu**
This pr is for issue #[4928 ](https://issues.apache.org/jira/browse/FLINK-4928) , it is based on #2808 for YarnResourceManager.

YARN Application Master Runner's responsibility is the following:
Read all configuration and environment variables, computing the effective configuration
Start all shared components (Rpc, HighAvailability Services)
Start the YarnResourceManager
Start the JobMaster
Handler the error of ResourceManager and JobMaster

### CODE REVIEW

**Comment by tillrohrmann** [comment]
Hi @shuai-xu, thanks for your contribution. Unfortunately, the PR does not compile correctly:
```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/apache/flink/flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java:[335,28] cannot find symbol
  symbol:   class YarnResourceManager
  location: class org.apache.flink.yarn.YarnFlinkApplicationMasterRunner
[INFO] 1 error
```

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+
+/**
+ * This class is the executable entry point for the YARN application master.
+ * It starts actor system and the actors for {@link org.apache.flink.runtime.jobmaster.JobMaster}
+ * and {@link org.apache.flink.yarn.YarnResourceManager}.
+ *
+ * The JobMasters handles Flink job execution, while the YarnResourceManager handles container
+ * allocation and failure detection.
+ */
+public class YarnFlinkApplicationMasterRunner implements LeaderContender, OnCompletionActions, FatalErrorHandler {
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java]: The current implementation tries to solve the Yarn cluster per job scenario, right? Maybe we could refactor the code such that we have `AbstractYarnFlinkApplicationMasterRunner` which does the generic Yarn initialization and two specializations `YarnPerJobFlinkApplicationMasterRunner` and `YarnClusterFlinkApplicationRunner`. The former basically works as this one. It starts a `ResourceManager` and a `JobManager` with the provided user code jar. The latter starts a `Dispatcher` (still to be implemented) and a `ResourceManager`. The `Dispatcher's` task is to receive job submission requests and to spawn for each request a new `JobMaster`. Then `JobMaster` is then responsible to run the submitted job.

**Comment by shuai-xu** [comment]
@tillrohrmann, thank for your review, the YarnResourceManager is in another pr   https://github.com/apache/flink/pull/2808, I will modify it according to your review soon.

**Comment by tillrohrmann** [comment]
Thanks for addressing my comments @shuai-xu. 

What you could do is to base one PR on the other PR that is not a problem. Then, the second PR can depend on the first one. But the first PR should be self-contained. Splitting them where each one of them has a dependency on the other is not ideal because then the tests cannot run through and merging one of them would render the flip-6 branch not compilable.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+					"ApplicationMaster hostname variable %s not set", Environment.NM_HOST.key());
+			LOG.info("YARN assigned hostname for application master: {}", appMasterHostname);
+
+			// try to start the rpc service
+			// using the port range definition from the config.
+			final String amPortRange = config.getString(
+					ConfigConstants.YARN_APPLICATION_MASTER_PORT,
+					ConfigConstants.DEFAULT_YARN_JOB_MANAGER_PORT);
+
+			synchronized (lock) {
```
Comment (shuai-xu) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java]: this is before starting, other thread has not been started,  so it seems no thead will call shutdown before it.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+	 * @param message The message for the runtime exception, with format variables as defined by
+	 *				{@link String#format(String, Object...)}.
+	 * @param values The format arguments.
+	 */
+	private static void require(boolean condition, String message, Object... values) {
+		if (!condition) {
+			throw new RuntimeException(String.format(message, values));
+		}
+	}
+	protected RpcService createRpcService(
```
Comment (shuai-xu) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java]: The RpcServiceUtils.createRpcService has to use a specified port, can not use port range.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	//  Utilities
+	// ------------------------------------------------------------------------
+
+	protected RpcService createRpcService(
+			Configuration configuration,
+			String bindAddress,
+			String portRange) throws Exception{
+		ActorSystem actorSystem = BootstrapTools.startActorSystem(configuration, bindAddress, portRange, LOG);
+		FiniteDuration duration = AkkaUtils.getTimeout(configuration);
+		return new AkkaRpcService(actorSystem, Time.of(duration.length(), duration.unit()));
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java]: Maybe you could extend `RcpServiceUtils.createRpcService` to also support port ranges.

Comment (StephanEwen) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java]: I think that would be a good idea.

**Comment by StephanEwen** [comment]
Manually merged to the FLIP-6 branch.

**Pull request closed without merging**